Fix markdown for the type system devdocs

ararslan opened the pull request:
One of the code blocks had a missing triple-backquote to close the block, which caused runaway formatting.

Another code block has been modified to use the backquote syntax both to enable Julia-specific syntax highlighting and for consistency with the rest of the document.

Fixes #20093

ararslan committed:
Fix markdown for the type system devdocs

One of the code blocks had a missing triple-backquote to close the
block, which caused runaway formatting.

Another code block has been modified to use the backquote syntax
both to enable Julia-specific syntax highlighting and for consistency
with the rest of the document.

Fixes #20093

tkelman commented on doc/src/devdocs/types.md:
````diff
@@ -100,6 +100,7 @@ julia> dump(Array)
          lb: Core.BottomType Union{}
          ub: Any
        body: Array{T,N} <: DenseArray{T,N}
+```
````
should this block be a doctest?

ararslan replied:
Dunno, none of the blocks in here are doctests. Converting some of them likely wouldn't hurt anything I would think. @JeffBezanson?

ararslan replied:
I've added a commit that converts many of the blocks to docstrings.

ararslan committed:
Convert several code blocks to doctests

Only a select few blocks have been converted. In particular, blocks
with pointers have been omitted, as have blocks with shortened REPL
output. This more conservative approach is intended to broaden the
tests for the new type system but also avoid any potential platform-
specific (or otherwise pesky) test failures.
